### PR TITLE
Request for Explicit Interop Test Suites from Mozilla

### PR DESCRIPTION
**Pull request by mprorock**
per email on the AC list

***
<a href="https://pr-preview.s3.amazonaws.com/mprorock/vc-wg-charter/pull/111.html" title="Last updated on Jun 5, 2022, 2:56 PM UTC (cce5faf)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-wg-charter/111/7589218...mprorock:cce5faf.html" title="Last updated on Jun 5, 2022, 2:56 PM UTC (cce5faf)">Diff</a>

### CODE REVIEW

**Comment by iherman** [comment]
cc @tantek

**Comment by clehner** [comment]
Public copy of the request: https://lists.w3.org/Archives/Public/public-new-work/2022Jun/0000.html

File: index.html [hunk 1]
```diff
@@ -418,6 +418,13 @@ <h2>Success Criteria</h2>
         <p>
           There should be testing plans for each specification, starting from the earliest drafts.
         </p>
+	<p>
+	In order to advance to Proposed Recommendation, each normative specification must have an open test suite
+	of every feature defined in the specification.
```
Comment (brentzundel) [inline index.html]:
```suggestion
	that tests every feature defined in the specification.
```

File: index.html [hunk 2]
```diff
@@ -418,6 +418,13 @@ <h2>Success Criteria</h2>
...
+	Additionally, two or more implementaitons for each feature should provide the ability to verify interoperability
```
Comment (brentzundel) [inline index.html]:
```suggestion
	Additionally, two or more implementaitons should demonstrate interoperability for each feature
```

File: index.html [hunk 3]
```diff
@@ -418,6 +418,13 @@ <h2>Success Criteria</h2>
...
+	by passing these open test suites, and by demonstrating that these implementations may interoperate with each other.
```
Comment (brentzundel) [inline index.html]:
```suggestion
	by passing these open test suites, and by showing that these implementations may interoperate with each other. 
```

File: index.html [hunk 4]
```diff
@@ -418,6 +418,13 @@ <h2>Success Criteria</h2>
         <p>
           There should be testing plans for each specification, starting from the earliest drafts.
         </p>
+	<p>
+	In order to advance to Proposed Recommendation, each normative specification must have an open test suite
+	of every feature defined in the specification.
+	Additionally, two or more implementaitons for each feature should provide the ability to verify interoperability
+	by passing these open test suites, and by demonstrating that these implementations may interoperate with each other. 
+	For example Verifiable Credentials produced by implementations must be consumed by other implementations and behave as specified. 
+	</p>
```
Comment (brentzundel) [inline index.html:427]:
```suggestion
	</p>
	<p>
A feature should be understood as a property or set of properties that are normatively required.
	</p>
```

**Review by brentzundel — CHANGES_REQUESTED** [review]
I've made some suggested changes for better clarity.

**Comment by brentzundel** [comment]
@mprorock if you accept my suggestions I will approve this PR, thank you for raising it.

**Comment by mprorock** [comment]
> @mprorock if you accept my suggestions I will approve this PR, thank you for raising it.

Accepted - thanks for the extra eyes!

**Review by brentzundel — APPROVED** [review]

File: index.html [hunk 5]
```diff
@@ -418,6 +418,16 @@ <h2>Success Criteria</h2>
         <p>
           There should be testing plans for each specification, starting from the earliest drafts.
         </p>
+	<p>
+	In order to advance to Proposed Recommendation, each normative specification must have an open test suite
+	that tests every feature defined in the specification.
+	Additionally, two or more implementaitons should demonstrate interoperability for each feature
```
Comment (Sakurann) [inline index.html]:
```suggestion
	Additionally, two or more normative implementations should demonstrate interoperability for each feature
```

**Review by Sakurann — APPROVED** [review]
please fix the typo in `implementations` before merging. Thank you for the PR

File: index.html (hunk 5 above)
Comment (brentzundel) [reply index.html]: I agree with the spelling fix, but am not totally sure what the difference is between an implementation and a normative implementation.

Comment (iherman) [reply index.html]: I would have had the same question... We do not really have the term "normative implementation". VC-s are normative data structures/formats whose exchange between implementations must be demonstrated; we do not specify the implementations themselves.

**Review by iherman — APPROVED** [review]

**Comment by brentzundel** [comment]
@msporny as you probably have the most direct experience with the current VC test suite, I am very interested in your opinion of these proposed changes.

File: index.html (hunk 5 above)
Comment (mprorock) [inline index.html]:
```suggestion
	Additionally, two or more implementations should demonstrate interoperability for each feature
```

**Comment by mprorock** [comment]
> please fix the typo in `implementations` before merging. Thank you for the PR

corrected

**Review by tantek — APPROVED** [review]
LGTM and this would resolve the specific issue that Mozilla's Formal Objection was based on. Thanks for putting in this work!

**Comment by msporny** [comment]
> @msporny as you probably have the most direct experience with the current VC test suite, I am very interested in your opinion of these proposed changes.

The more modern test suites that we're planning on using for the VC 2.0 work will be able to test interoperability in the way this PR is suggesting. 

The only question I have is whether or not the test suites can utilize protocols to demonstrate interoperability? Given that test suites are non-normative, I suggest the answer to this question is "Yes, the test suite can utilize VC protocols to demonstrate interoperability." For example, the credential refresh and credential status require protocols to be used. I believe this would address Google and Mozilla's concerns, while not crossing the hard line that Microsoft has about the WG not working on protocols.

IOW, one way through this is to say the WG will not normatively define protocols, but can utilize protocols to demonstrate interoperability on VC 2.0 data model features. This all keeps going back to whether or not the VC2WG is generating data model specifications or data model + protocol specifications. I believe we have consensus to work on data models normatively and protocols non-normatively... and we can prove interoperability by using a combination of normative specification text and non-normative test suites (that utilize particular protocols).

**Comment by iherman** [comment]
@msporny I do not think there is a problem. The text in the charter does not say how the interchange happens, that can be done via protocol agreed upon by parties, emails, or by snail mail...

**Comment by iherman** [comment]
I will start the formal process by W3M for the approval of the charter.

**Comment by iherman** [comment]
> LGTM and this would resolve the specific issue that Mozilla's Formal Objection was based on. Thanks for putting in this work!

Thanks Tantek.

**Pull request merged**